File: 279-perfect-squares/279-perfect-squares.cpp
class Solution {
public:
    typedef unsigned long long ull;
    int numSquares(int n) {
        vector<ull> sq_nums;
        for (int i = 0; i < sqrt(n) + 1; ++i) sq_nums.push_back(i * i);
        vector<ull> dp(n + 1, INT_MAX);
        dp[0] = 0;
        
        for (ull& num: sq_nums) {
            for (int i = num; i <= n; ++i) {
                dp[i] = min(dp[i], dp[i - num] + 1);
            }
        }
        return dp[n];
    }
};

File: 279-perfect-squares/279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        sq_nums = [i ** 2 for i in range(int(n ** 0.5 + 1))]
        dp = [float('inf')] * (n + 1)
        dp[0] = 0
        
        for i in range(1, n + 1):
            for sq_num in sq_nums:
                if i >= sq_num:
                    dp[i] = min(dp[i], dp[i - sq_num] + 1)
        return dp[-1]